### ISSUE TITLE
File Upload Service Python 3 - allow for extra data/dynamic dir placement

### ISSUE THREAD

**Issue by singlecheeze** [crossbario/crossbar#505]
The resumable.js spec states that extra data may be appended to the query config option:

> query Extra parameters to include in the multipart POST with data. This can be an object or a function. If a function, it will be passed a ResumableFile object (Default: {})

Crossbar already uses this config parameter for json data that hooks up to `on_progress` and `session` as such:

``` javascript
r = new Resumable({
            target: 'upload',
            chunkSize: 1024 * 512, // Deault 1MB 1024 * 1024
            forceChunkSize: true, // https://github.com/23/resumable.js/issues/51
            simultaneousUploads: 4,
            testChunks: true,
            query: {
                on_progress: 'mono.upload.progress',
                session: sess.id
            }
        });
```

What I am proposing is allowing the user some functionality to expand on this further. What I would like to use it for is dynamically assigning a destination directory for the final upload that is a child of the crossbar configuration directory `"directory": "../uploaded"`

So in use a user could do this:

``` javascript
            query: {
                on_progress: 'mono.upload.progress',
                session: sess.id,
                child_dir: 'this_is_a_child_dir'
            }
```

And the result on successful file upload in crossbar would be before the final move occurs:

``` python
if query["child_dir"]:
     """ Maybe check to see if the target child dir is in fact a child 
          if full path of child is already given in the option, 
          this might be a way of adding nesting at n levels as well """
     destination = os.path.join("<<directory from directory config option>>", "<<child_dir>>")
```

Without this option, all files uploaded get put in exactly one directory on the server without the ability to add sub dirs even at one level deep since crossbar is a "Black Box" to a client application.

This is only an example of an implementation, if someone has something better or a way of already accomplishing this (I couldn't find one from the docs) then please share.

The only way to get around this currently is to wait for a `r.on('fileSuccess', function (file, message)` on the client and then make the client responsible to call a method to move the file to it's final destination. Having another mechanism on the server listen to `on_progress` and do something then will not solve this as there is still no way to pass additional data to that method on the server side.

**Comment by markope** [comment]
hm, making subdirs one level deep sounds like a very specific requirement. I feel a solution to this should be more generic in some way. Could you tell me a bit about the actual use case you are after?

**Comment by singlecheeze** [comment]
Use case is just to have the upload data contain a target directory if
specified in the resumable object and to place the reassembled resumable
pieces there on successful upload vs a singular target directory for all
uploads.

This would be awesome Marko for the specific case where a file actually
needs to be resumed. The only way to achieve this currently is in the
success event handler on the resumable object. I have to make a seperate
call to actually move the file to the directory on the server. If the
resumable object could cache this target directory, then in the case of the
network breaking or the app breaking, on an actual resume operation of some
sort, the target dir wouldnt be lost like it is currently in the case of my
seperate call to move the file. The "state" of my seperate call is lost
currently as it is not directly tied to the resumable object. Essentially
what would be required is moving the call to move the file onto the server
vs the way i have it now on the client so it is not lost.

Let me know if that makes sense. If not I'll try to explain better.

**Comment by markope** [comment]
So if I get you right you have an upload initiating client session and potentially a different upload finishing client session. The initiating client session should tell in what subdir the file should eventually be moved and the server should keep that in mind accross the session switch. 

>Question: Why not have the finishing session always tell where to move the file?


**more thoughts:**
Also if the purpose is to seperate upload areas for different users this is not a secure approach. Users would be able to do a hack to upload into other users folders. In fact I have been working on a secure separation feature seperating uploads by the auth_ids of the sessions into different folders. This however presumes you are using the crossbar authentication methods and don't just do anonymous connections.

Another more generic solution may be to specify a post processing call (i.e. a URI and a payload) in the extra params of the resumable. Those would be called _on-finish_ so nobody has to listen to progress or watch folders on server side. This would also cover a wider area of possible use cases than moving a file to a folder. (like e.g. decompress, process images, etc...) 
The problems with calls issued by crossbar however is that crossbar will ignore any errors on return of the call. That call wouldn't have any more purpose then than the already existing publish.  

In general I think post-processing uploads should stay out of scope of crossbar. 

Your comments on this would be welcome. 
@oberstet what do you think.

cheers,
Marko

**Comment by singlecheeze** [comment]
@markope I think this is the most flexible solution:
> Another more generic solution may be to specify a post processing call (i.e. a URI and a payload) in the extra params of the resumable.

With the expressed caveat that this call will originate on the on-finish of the server, not the client.

Watching folders is yet another thing I had to do to get around this without having to listen to progress events from the client side like you state. BTW this is a great package if we wanted to get DirWatcher features back into Py 3 Crossbar @oberstet : https://pypi.python.org/pypi/watchdog

In regard to this:
> The problems with calls issued by crossbar however is that crossbar can not do custom error handling on any errors on return of the call. That call wouldn't have any more purpose then than the already existing publish.

I respectfully disagree, a file upload capability without the ability to do anything with the file from the server side negates it's usefulness. Normally in another web framework, a developer would have direct access to these functions, but since Crossbar is essentially a "black box" (And it should be, as the decoupled nature and all), we must allow for further control of this feature else make it external to the solution completely... just my $.02

**Comment by markope** [comment]
The scenario I had in mind when I said _publish is the same as call without return value_ is to use autobahn-python on the server side and subscribe to the progress event in the backend. Would this not be a good solution for your use case?

Regarding the usefulness of a pure file uploader I think using watchdog or listening to finish events is good practice. In any case I do agree that the finish event/call should be enrichable with custom data, be it a call or a publish.

@oberstet Do you also prefer an additional rpc call for post processing?

To summarize: 
My main argument against an additional call is the additional API elements while having little benefit over the publish.
The only argument in favor of an rpc call I understood is that it feels more explicit when implementing. Help me out here if I missed something.

**Comment by oberstet** [comment]
The only advantage I could imagine with having a procedure call (in addition to the event published) could be: user feedback rgd any postprocessing performed. But that could also be done by publishing a topic in the uploaded event, to which later postprocessing events would be published. Not sure if its worth though.

Am 20.11.2015 9:30 nachm. schrieb Marko Petzold <notifications@github.com>:

The scenario I had in mind when I said publish is the same as call without return value is to use autobahn-python on the server side and subscribe to the progress event in the backend. Would this not be a good solution for your use case?

Regarding the usefulness of a pure file uploader I think using watchdog or listening to finish events is good practice. In any case I do agree that the finish event/call should be enrichable with custom data, be it a call or a publish.

@oberstet<https://github.com/oberstet> Do you also prefer an additional rpc call for post processing?

To summarize:
My main argument against an additional call is the additional API elements while having little benefit over the publish.
The only argument in favor of an rpc call I understood is that it feels more explicit when implementing. Help me out here if I missed something.

**Comment by singlecheeze** [comment]
Yes true that would work fine and move the post processing to the server
side vs counting on a client to kick something off after upload
completes... but this still doesn't allow some kind of data payload to go
with the file.

Originally the people who made resumable.js allowed for extra data to be
sent with the file upload. But since the way the upload service in crossbar
is configured and then set and crossbar runs as a black box, there is no
way to send any data with the file upload.

At least creating a custom property on the resumable object in the extra
data session where we already have session data that is passed through the
uploader and on to the progress object would work as that could be read if
set and then some post processing could be done on the upload. Right now I
can upload files, but once uploaded, I have no meta data of what the file
is or where it should go.

Your solution would work great and will be the path I take if we can just
get a property on the resumable object that say accepts json data and is in
turn passed to that uploads progress topic, that once file upload
completes, can be used to manipulate the post processing.

**Comment by oberstet** [comment]
> Do you also prefer an additional rpc call for post processing?

No, please leave it as is for now

**Comment by markope** [comment]
I implemented a feature to add extra info to the resumable.js post that is again published by crossbar to any listening client. See my updated fileupload example in the crossbarexamples repo.

**Comment by singlecheeze** [comment]
Marko I know you said you have been busy lately so I really appreciate you
making time for this today!

**Comment by singlecheeze** [comment]
Hey @markope I'm lookin around in the repo for the example but I must be blind. Can you please link?

**Comment by markope** [comment]
maybe they didnt merge the examples repo yet. you can take a look into my personal fork of crossbarexamples repo.

**Comment by singlecheeze** [comment]
@markope 

Marko.... My sincere apologizes. I think I missed the mark on my ask here. Please do not remove any of the functionality that you have added as it is very very good. My mistkae, after reviewing the functionality you have added, was the fact that I need extra data to be on the resumable file object, not on the resumable object itself.

I did a test case where I create a resumable object (below), and then try and change the finish_extra on the resumable object before I begin a file upload, but as I suspected, this sets the property of the last file uploaded. What I need is extra data that is cached with the file... pretty much exactly what you have implemented for the resumable object, but at the file level of granularity.

```js
r = new Resumable({
            target: 'upload',
            chunkSize: 1024 * 512, // Deault 1MB 1024 * 1024
            forceChunkSize: true, // https://github.com/23/resumable.js/issues/51
            simultaneousUploads: 4,
            testChunks: true,
            query: {
                on_progress: 'mono.upload.on_progress',
                session: sess.id //, 
                // ****************** Below in the fileAdded event I will adjust query opts
                //chunk_extra: JSON.stringify({
                //    SR: currentSR,
                //    extra_info2: "blah"
                //}),
                //finish_extra: JSON.stringify({
                //    SR: currentSR,
                //    extra_info2: "blah"
                //})
            }
        });

r.on('fileAdded', function (file) {
                if (currentSR) {
                    //Set file SR to current SR so that if user browses to new SR
                    //file will still be uploaded correctly under it's cached SR
                    file.resumableObj.opts.query.finish_extra = JSON.stringify({
                        SR: currentSR,
                        extra_info2: "blah"
                    });

                    r.upload();
                } else {
                    console.log("No SR Selected!");
                }
            });
```

**Comment by singlecheeze** [comment]
Bumping this one @markope. You have done a tremendous job on this so far and I know many people will use the addition of data to the file object. Any chance you have a few minutes to work on this? Any questions? I want to help any way I can.

**Comment by markope** [comment]
I guess I can't do anything about this. If the POST request of resumable.js includes extra data in the content body as specified, then the crossbar fileupload feature is going to use it and send it back in the progress message. 
You can either ask the resumable.js guys to add this to their POST request or create a POST request on your own.

**Comment by markope** [comment]
btw: did the last fix fix your problem with the makedir?

**Comment by singlecheeze** [comment]
@markope finally got time today to test out your latest update...

Mysteriously I still got the below error... and in fact not chunk_1 exists in the file system...

```bash
  File "/usr/local/lib/python3.5/site-packages/twisted/web/server.py", line 183, in process
    self.render(resrc)
  File "/usr/local/lib/python3.5/site-packages/twisted/web/server.py", line 234, in render
    body = resrc.render(self)
  File "/usr/local/lib/python3.5/site-packages/twisted/web/resource.py", line 250, in render
    return m(request)
  File "/usr/local/lib/python3.5/site-packages/crossbar/twisted/resource.py", line 439, in render_POST
    with open(fileTempDir.child('chunk_' + str(cn)).path, 'rb') as ff:
builtins.FileNotFoundError: [Errno 2] No such file or directory: '/home/dave/monocle/frontend/temp/APM00113301408_SPA_2015-12-02_15-24-35_2df0a6_data.zip/chunk_1'

```

**Comment by singlecheeze** [comment]
![screenshot from 2015-12-08 20 56 22](https://cloud.githubusercontent.com/assets/12156213/11674369/4343f95c-9dee-11e5-8a32-1d889f713382.png)

**Comment by markope** [comment]
Dave, thanks for your patience with me. I hope you don't get annoyed by these constant fails. I really tried to reproduce the error in many ways, but did not succeed. In my setup all goes well. Do you have any idea how you could help me point to the error?

**Comment by singlecheeze** [comment]
@markope No patience needed, this is a very complex issue and I really appreciate the help!

Testing some other items I got this more robust error tonight:
```text
2015-12-09T19:39:35-0500 [Router      10834] Unhandled Error
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/twisted/web/http.py", line 1445, in dataReceived
    finishCallback(data[contentLength:])
  File "/usr/local/lib/python3.5/site-packages/twisted/web/http.py", line 1703, in _finishRequestBody
    self.allContentReceived()
  File "/usr/local/lib/python3.5/site-packages/twisted/web/http.py", line 1766, in allContentReceived
    req.requestReceived(command, path, version)
  File "/usr/local/lib/python3.5/site-packages/twisted/web/http.py", line 832, in requestReceived
    self.process()
--- <exception caught here> ---
  File "/usr/local/lib/python3.5/site-packages/twisted/web/server.py", line 183, in process
    self.render(resrc)
  File "/usr/local/lib/python3.5/site-packages/twisted/web/server.py", line 234, in render
    body = resrc.render(self)
  File "/usr/local/lib/python3.5/site-packages/twisted/web/resource.py", line 250, in render
    return m(request)
  File "/usr/local/lib/python3.5/site-packages/crossbar/twisted/resource.py", line 439, in render_POST
    with open(fileTempDir.child('chunk_' + str(cn)).path, 'rb') as ff:
builtins.FileNotFoundError: [Errno 2] No such file or directory: '/home/dave/monocle/frontend/temp/ubuntu-font-family-0.83.zip/chunk_2'
```
Then with another upload (different file) going from a different browser on another computer to the same router instance (funny enough some files I queue up complete successfully and some do not...):
```text
2015-12-09T19:43:20-0500 [Router      10834] Unhandled Error
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/twisted/web/http.py", line 1445, in dataReceived
    finishCallback(data[contentLength:])
  File "/usr/local/lib/python3.5/site-packages/twisted/web/http.py", line 1703, in _finishRequestBody
    self.allContentReceived()
  File "/usr/local/lib/python3.5/site-packages/twisted/web/http.py", line 1766, in allContentReceived
    req.requestReceived(command, path, version)
  File "/usr/local/lib/python3.5/site-packages/twisted/web/http.py", line 832, in requestReceived
    self.process()
--- <exception caught here> ---
  File "/usr/local/lib/python3.5/site-packages/twisted/web/server.py", line 183, in process
    self.render(resrc)
  File "/usr/local/lib/python3.5/site-packages/twisted/web/server.py", line 234, in render
    body = resrc.render(self)
  File "/usr/local/lib/python3.5/site-packages/twisted/web/resource.py", line 250, in render
    return m(request)
  File "/usr/local/lib/python3.5/site-packages/crossbar/twisted/resource.py", line 439, in render_POST
    with open(fileTempDir.child('chunk_' + str(cn)).path, 'rb') as ff:
builtins.FileNotFoundError: [Errno 2] No such file or directory: '/home/dave/monocle/frontend/temp/a5UA0000000L1rvMAC~1~WestervilleXIO-vbxbcl02psdu_2015_10_23_1029.tar.gz/chunk_1'
```

**Comment by singlecheeze** [comment]
@markope can you please check to make sure your latest updates to you repo are reflected in crossbar master? I just reinstalled crossbar from latest master and am running into the missing chunk_1 issue referenced above again... Didn't we fix this with #544 ?

**Comment by markope** [comment]
yes we did, but three days after the fix got merged another contributor removed this fix.
@hawkowl how should we proceed about this? Do you have an alternative way to fix this?

**Comment by singlecheeze** [comment]
Oh man... wonder why? It's def back:
```text
  File "/usr/local/lib/python3.5/dist-packages/twisted/web/server.py", line 234, in render
    body = resrc.render(self)
  File "/usr/local/lib/python3.5/dist-packages/twisted/web/resource.py", line 250, in render
    return m(request)
  File "/usr/local/lib/python3.5/dist-packages/crossbar/twisted/resource.py", line 481, in render_POST
    mergeFile()
  File "/usr/local/lib/python3.5/dist-packages/crossbar/twisted/resource.py", line 334, in mergeFile
    with open(fileTempDir.child('chunk_' + str(cn)).path, 'rb') as ff:
builtins.FileNotFoundError: [Errno 2] No such file or directory: '/home/dave/monocle/frontend/temp/APM00113301408_SPA_2015-12-02_15-24-35_2df0a6_data.zip/chunk_1'
```

**Comment by singlecheeze** [comment]
@markope @oberstet @hawkowl @meejah 
I REALLY appreciate how closely everyone keeps an eye on GitHub and comments! You are directly effecting the success of Crossbar!

**Comment by singlecheeze** [comment]
@markope @hawkowl @oberstet 
Why was the fix removed in the first place? Does it conflict with something? Why don't we just put in some six check to see if py2 vs py3?

Can we please have this fix merged in again quickly?

**Comment by oberstet** [comment]
Because there is no hurry and sprinkling py2/3 checks all over the place is not the way to go.

**Comment by oberstet** [comment]
I don't have time, but from a quick look, the code has multiple issues:

* Use of `FilePath`
* Use of string concatenation instead of interpolation.
* What's that "kfhf3kz412uru578e38viokbjhfvz4w__" magic constant. That must be solved differently.

**Comment by oberstet** [comment]
https://github.com/crossbario/crossbar/issues/584

**Comment by singlecheeze** [comment]
I think that this is an applicable use case for a check as it is directly related to a difference in the platforms. File upload on py 3 is broken without this and weighing having file upload vs. whatever this fix broke, I think is definitely worth is as file upload is an integral piece to most projects I would assume. 

That's just my bias .02...

I appreciate the work @markope has put into this functionality and he has been very transparent in the fact that the feature is not fully complete yet, though it would be nice if it _worked_ especially with all of the push to py 3

**Comment by oberstet** [comment]
> I think that this is an applicable use case for a check

No, it's not.

**Comment by singlecheeze** [comment]
:cry: :wink:

**Comment by singlecheeze** [comment]
Hey @markope the guys/gals from ResumableJS got back to me: 
https://github.com/23/resumable.js/issues/274

How do you think it would be best to publish file query data exclusive from Resumable object query data?

**Comment by markope** [comment]
Did you check that out? The fileupload in crossbar just reflects in the WAMP event whatever chunk_extra or finish_extra has been posted. So there is no change necessary in crossbar for this to work.

**Comment by oberstet** [comment]
the file upload service has been removed (for now)
